crates/sui-core/src/transaction_signing_filter.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use sui_config::transaction_deny_config::TransactionDenyConfig;
+use sui_types::{
+    error::{SuiError, SuiResult, UserInputError},
+    messages::{Command, InputObjectKind, TransactionData, TransactionDataAPI},
+    storage::BackingPackageStore,
+};
+
+#[cfg(test)]
+#[path = "unit_tests/transaction_deny_tests.rs"]
+mod transaction_deny_tests;
+
+macro_rules! deny_if_true {
+    ($cond:expr, $msg:expr) => {
+        if ($cond) {
+            return Err(SuiError::UserInputError {
+                error: UserInputError::TransactionDenied {
+                    error: $msg.to_string(),
+                },
+            });
+        }
+    };
+}
+
+/// Check that the provided transaction is allowed to be signed according to the
+/// deny config.
+pub fn check_transaction_for_signing(
+    tx_data: &TransactionData,
+    input_objects: &[InputObjectKind],
+    filter_config: &TransactionDenyConfig,
+    package_store: &impl BackingPackageStore,
+) -> SuiResult {
+    check_disabled_features(filter_config, tx_data)?;
+
+    check_sender_and_sponsor(filter_config, tx_data)?;
+
+    check_input_objects(filter_config, input_objects)?;
+
+    check_package_dependencies(filter_config, tx_data, package_store)?;
+
+    Ok(())
+}
+
+fn check_disabled_features(
+    filter_config: &TransactionDenyConfig,
+    tx_data: &TransactionData,
+) -> SuiResult {
+    deny_if_true!(
+        filter_config.user_transaction_disabled(),
+        "Transaction signing is temporarily disabled"
+    );
+
+    if !filter_config.package_publish_disabled() && !filter_config.package_upgrade_disabled() {
+        return Ok(());
+    }
+    for command in tx_data.kind().iter_commands() {
```

Comment:
Does it make sense to add a facility to disable each command? (e.g. also `Split`, `Merge` and `Transfer`?). It's quite an extreme move, and I can't think of an immediate reason to use it, but I'd hate to need it and not have it.